h/static/scripts/controllers/form-input-controller.js

```diff
@@ -0,0 +1,32 @@
+'use strict';
+
+var inherits = require('inherits');
+
+var Controller = require('../base/controller');
+var setElementState = require('../util/dom').setElementState;
+
+/**
+ * Controller for form input fields.
+ */
+function FormInputController(element) {
```

Comment:
In https://github.com/hypothesis/h/pull/3815 I make use of this encapsulation to map different native events to the same synthetic event depending on input type.

A caveat specific to focus events is that `focus` and `blur` do not bubble, although you can intercept them before they actually trigger in the capture phase.